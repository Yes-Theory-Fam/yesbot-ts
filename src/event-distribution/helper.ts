import {
  CommandHandler,
  DiscordEvent,
  HandlerInfo,
  MessageRelatedOptions,
} from ".";
import {
  HIOC,
  InstanceOrConstructor,
  StringIndexedHIOCTree,
} from "./types/hioc";
import {
  GuildChannel,
  GuildMember,
  Message,
  PartialMessage,
  TextBasedChannel,
} from "discord.js";

export const getIdFromCategoryName = (name: string) =>
  `c_${name.toLowerCase()}`;

export const getIocName = <T extends DiscordEvent>(
  ioc: InstanceOrConstructor<CommandHandler<T>>
) => {
  return typeof ioc === "function" ? ioc.name : ioc.constructor.name;
};

export const collectChannelDefinitions = (options: MessageRelatedOptions) => {
  const channels = options.channelNames ?? [];
  const categories = options.categoryNames ?? [];
  if (channels.length === 0 && categories.length === 0) channels.push("");

  return [...channels, ...categories.map((c) => getIdFromCategoryName(c))];
};

type HandlerKeyFromChannelIdResolver = (channelIdentifier: string) => string[];

export const withMessageRelatedInfo = (
  message: Message | PartialMessage,
  member: GuildMember | null,
  resolver: HandlerKeyFromChannelIdResolver
): HandlerInfo[] => {
  const getChannelIdentifier = (channel: TextBasedChannel) =>
    channel.type === "DM" ? channel.id : channel.name;

  const channel = message.channel;
  const channelIdentifier = getChannelIdentifier(channel);

  const baseInfo = {
    member: member,
    isDirectMessage: message.channel.type === "DM",
  };

  const info = [
    {
      ...baseInfo,
      handlerKeys: resolver(channelIdentifier),
    },
  ];

  const maybeCategory = (channel as GuildChannel).parent;
  if (maybeCategory) {
    const normalizedCategoryName = maybeCategory.name
      .match(/[a-z\d\s.]+/gi)[0]
      .trim();
    const categoryIdentifier = getIdFromCategoryName(normalizedCategoryName);

    info.push({
      ...baseInfo,
      handlerKeys: resolver(categoryIdentifier),
    });
  }

  return info;
};

export const addToTree = <T extends DiscordEvent>(
  keys: string[],
  hioc: HIOC<T>,
  tree: StringIndexedHIOCTree<T>
) => {
  if (keys.length === 1) {
    const [key] = keys;
    tree[key] ??= [];
    const handlers = tree[key] as HIOC<T>[];

    handlers.push(hioc);
    return;
  }

  const [key, ...rest] = keys;

  const subTree = tree[key];
  if (Array.isArray(subTree)) {
    tree[key] = { "": subTree };
  }

  tree[key] ??= {};
  addToTree(rest, hioc, tree[key] as StringIndexedHIOCTree<T>);
};
