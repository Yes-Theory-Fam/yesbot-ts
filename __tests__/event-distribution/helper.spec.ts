import { addToTree } from "../../src/event-distribution/helper.js";
import {
  HIOC,
  StringIndexedHIOCTree,
} from "../../src/event-distribution/types/hioc.js";
import { DiscordEvent } from "../../src/event-distribution/index.js";

describe("Event Distribution helper", () => {
  it("Adds a single key event to the tree", () => {
    const hioc = {} as HIOC<DiscordEvent.READY>;
    const tree = {} as StringIndexedHIOCTree<DiscordEvent.READY>;

    addToTree([""], hioc, tree);

    expect(tree[""]).toStrictEqual([hioc]);
  });

  it("Adds a two key event to the tree", () => {
    const hioc = {} as HIOC<DiscordEvent.REACTION_ADD>;
    const tree = {} as StringIndexedHIOCTree<DiscordEvent.REACTION_ADD>;

    addToTree(["first", "second"], hioc, tree);

    expect(tree).toMatchSnapshot();
  });

  it("Adds a three key event to the tree", () => {
    const hioc = {} as HIOC<DiscordEvent.MESSAGE>;
    const tree = {} as StringIndexedHIOCTree<DiscordEvent.MESSAGE>;

    addToTree(["channel", "trigger", "subTrigger"], hioc, tree);

    expect(tree).toMatchSnapshot();
  });

  it("Converts an existing array to an object", () => {
    const hioc = {} as HIOC<DiscordEvent.MESSAGE>;
    const tree = {} as StringIndexedHIOCTree<DiscordEvent.MESSAGE>;

    addToTree(["first"], hioc, tree);

    expect(tree["first"]).toStrictEqual([hioc]);

    addToTree(["first", "second"], hioc, tree);
    expect(Array.isArray(tree["first"])).toBe(false);

    const firstTree = tree[
      "first"
    ] as StringIndexedHIOCTree<DiscordEvent.MESSAGE>;

    expect(firstTree[""]).toBeTruthy();
    expect(firstTree["second"]).toBeTruthy();
  });

  it("Works with a bunch of combinations", () => {
    const hioc = {} as HIOC<DiscordEvent.MESSAGE>;
    const tree = {} as StringIndexedHIOCTree<DiscordEvent.MESSAGE>;

    addToTree(["first"], hioc, tree);
    addToTree(["second"], hioc, tree);
    addToTree(["third", "example", "sub"], hioc, tree);
    addToTree(["first", "second"], hioc, tree);
    addToTree(["first", "third"], hioc, tree);
    addToTree(["first", "third"], hioc, tree);

    expect(tree).toMatchSnapshot();
  });
});
