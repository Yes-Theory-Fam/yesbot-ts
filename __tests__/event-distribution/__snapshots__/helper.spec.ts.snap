// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Event Distribution helper Adds a three key event to the tree 1`] = `
{
  "channel": {
    "trigger": {
      "subTrigger": [
        {},
      ],
    },
  },
}
`;

exports[`Event Distribution helper Adds a two key event to the tree 1`] = `
{
  "first": {
    "second": [
      {},
    ],
  },
}
`;

exports[`Event Distribution helper Works with a bunch of combinations 1`] = `
{
  "first": {
    "": [
      {},
    ],
    "second": [
      {},
    ],
    "third": [
      {},
      {},
    ],
  },
  "second": [
    {},
  ],
  "third": {
    "example": {
      "sub": [
        {},
      ],
    },
  },
}
`;
